### PR TITLE
feat: assign spoke pool events to bundles

### PR DESCRIPTION
**Pull request by melisaguevara**
This PR adds the functionality to reconstruct executed bundles in order to identify the events that were part of it and save them in the database. It also uses the stored events to identify refunds of expired deposits.

### CODE REVIEW

File: packages/indexer/src/database/BundleRepository.ts [hunk 1]
```diff
@@ -371,4 +380,166 @@ export class BundleRepository extends utils.BaseRepository {
...
+  ) {
+    const eventsRepo = this.postgres.getRepository(entities.BundleEvents);
+
+    // Store bundle deposits
+    const formattedDeposits = this.formatBundleEvents(
+      entities.BundleEventTypes.Deposit,
+      bundleData.bundleDepositsV3,
+      bundleId,
+    );
+    const storedDeposits = await eventsRepo.insert(formattedDeposits);
```
Comment (amateima) [inline packages/indexer/src/database/BundleRepository.ts:450]: All these inserts should be split in multiple chunks with fixed length

File: packages/indexer/src/utils/bundleBuilderUtils.ts [hunk 2]
```diff
@@ -1,7 +1,8 @@
+import { utils, clients } from "@across-protocol/sdk";
+import { LoadDataReturnValue } from "@across-protocol/sdk/dist/cjs/interfaces/BundleData";
```
Comment (amateima) [inline packages/indexer/src/utils/bundleBuilderUtils.ts:2]: Can the type be imported only from the `/dist/cjs`?

File: packages/indexer/src/utils/bundleBuilderUtils.ts [hunk 3]
```diff
@@ -57,6 +58,27 @@ export async function resolveMostRecentProposedAndExecutedBundles(
   return { lastExecutedBundle, lastProposedBundle };
 }
 
+/**
+ * Given a bundle entitie with its related ranges and proposals,
```
Comment (amateima) [inline packages/indexer/src/utils/bundleBuilderUtils.ts:62]:
```suggestion
 * Given a bundle entity with its related ranges and proposals,
```

nit

File: packages/indexer/src/web3/constants.ts [hunk 4]
```diff
@@ -91,3 +91,36 @@ export function getMaxBlockLookBack(chainId: number) {
...
+  [CHAIN_IDs.LINEA]: 3,
+  [CHAIN_IDs.LISK]: 2,
+  [CHAIN_IDs.MAINNET]: 12,
+  [CHAIN_IDs.MODE]: 2,
+  [CHAIN_IDs.OPTIMISM]: 2,
+  [CHAIN_IDs.POLYGON]: 2,
+  [CHAIN_IDs.REDSTONE]: 2,
+  [CHAIN_IDs.SCROLL]: 3,
+  [CHAIN_IDs.ZK_SYNC]: 1,
+  [CHAIN_IDs.ZORA]: 2,
```
Comment (amateima) [inline packages/indexer/src/web3/constants.ts:110]: Should we add world chain too?

File: packages/indexer/src/web3/constants.ts [hunk 5]
```diff
@@ -91,3 +91,36 @@ export function getMaxBlockLookBack(chainId: number) {
...
+ * Resolves the block time in seconds for a given chain id
+ * @param chainId Chain id to resolve block time for
+ * @returns The average block time in seconds from {@link BLOCK_TIME_SECONDS} or a default of 10
+ */
+export function getBlockTime(chainId: number) {
+  const blockTime = BLOCK_TIME_SECONDS[chainId];
+
+  if (!blockTime) {
+    return 10;
+  }
```
Comment (amateima) [inline packages/indexer/src/web3/constants.ts:123]: Should we return a value or throw an error to surface the missing value?

File: packages/indexer-database/src/entities/BundleEvents.ts [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+  Deposit = "deposit",
+  ExpiredDeposit = "expiredDeposit",
+  Fill = "fill",
+  SlowFill = "slowFill",
+  UnexecutableSlowFill = "unexecutableSlowFill",
+}
+
+@Entity()
+@Unique("UK_bundleEvents_eventType_relayHash", ["eventType", "relayHash"])
+export class BundleEvents {
```
[the remark this answers is not in the document]
Comment (melisaguevara) [reply packages/indexer-database/src/entities/BundleEvents.ts]: Renamed here [09c466a](https://github.com/across-protocol/indexer/pull/77/commits/09c466aadfb787bdcb1300ff62aed21784d26ccd)

File: packages/indexer-database/src/entities/BundleEvents.ts [hunk 7]
```diff
@@ -0,0 +1,38 @@
+import {
+  Column,
+  Entity,
+  ManyToOne,
+  PrimaryGeneratedColumn,
+  Unique,
+} from "typeorm";
+import { Bundle } from "./Bundle";
+
+export enum BundleEventTypes {
```
[the remark this answers is not in the document]
Comment (melisaguevara) [reply packages/indexer-database/src/entities/BundleEvents.ts]: Renamed here too [09c466a](https://github.com/across-protocol/indexer/pull/77/commits/09c466aadfb787bdcb1300ff62aed21784d26ccd)

File: packages/indexer-database/src/entities/BundleEvents.ts [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+  id: number;
+
+  @ManyToOne(() => Bundle, (bundle) => bundle.events)
+  bundle: Bundle;
+
+  @Column()
+  bundleId: number;
+
+  @Column({ type: "enum", enum: BundleEventTypes })
+  eventType: BundleEventTypes;
```
[the remark this answers is not in the document]
Comment (melisaguevara) [reply packages/indexer-database/src/entities/BundleEvents.ts]: This was also renamed here [09c466a](https://github.com/across-protocol/indexer/pull/77/commits/09c466aadfb787bdcb1300ff62aed21784d26ccd)

File: packages/indexer/src/database/BundleRepository.ts [hunk 9]
```diff
@@ -1,4 +1,13 @@
 import winston from "winston";
+import {
+  BundleDepositsV3,
+  BundleExcessSlowFills,
+  BundleFillsV3,
+  BundleSlowFills,
+  ExpiredDepositsToRefundV3,
+  LoadDataReturnValue,
+} from "@across-protocol/sdk/dist/cjs/interfaces/BundleData";
+import { getRelayHashFromEvent } from "@across-protocol/sdk/dist/cjs/utils/SpokeUtils";
```
[the remark this answers is not in the document]
Comment (melisaguevara) [reply packages/indexer/src/database/BundleRepository.ts]: Fixed here [1747f18](https://github.com/across-protocol/indexer/pull/77/commits/1747f18bcfc822a1dd6896b6d67001d0767e5ccd)

File: packages/indexer/src/database/BundleRepository.ts [hunk 10]
```diff
@@ -371,4 +380,166 @@ export class BundleRepository extends utils.BaseRepository {
...
+    eventsType: entities.BundleEventTypes,
+    bundleEvents:
+      | BundleDepositsV3
+      | BundleSlowFills
+      | BundleExcessSlowFills
+      | ExpiredDepositsToRefundV3,
+    bundleId: number,
+  ): {
+    bundleId: number;
+    relayHash: any;
```
[the remark this answers is not in the document]
Comment (melisaguevara) [reply packages/indexer/src/database/BundleRepository.ts]: oops! Fixed here [09c466a](https://github.com/across-protocol/indexer/pull/77/commits/09c466aadfb787bdcb1300ff62aed21784d26ccd)

File: packages/indexer/src/database/BundleRepository.ts (hunk 1 above)
Comment (melisaguevara) [reply packages/indexer/src/database/BundleRepository.ts]: Good call! Done here [670701c](https://github.com/across-protocol/indexer/pull/77/commits/670701cf24a54e4fc1874a4cf539a93d2b718264)

File: packages/indexer/src/utils/bundleBuilderUtils.ts (hunk 2 above)
Comment (melisaguevara) [reply packages/indexer/src/utils/bundleBuilderUtils.ts]: Fixed here [1747f18](https://github.com/across-protocol/indexer/pull/77/commits/1747f18bcfc822a1dd6896b6d67001d0767e5ccd)

File: packages/indexer/src/web3/constants.ts (hunk 4 above)
Comment (melisaguevara) [reply packages/indexer/src/web3/constants.ts:115]: [fc62a98](https://github.com/across-protocol/indexer/pull/77/commits/fc62a984407bd7856ff3bf98f42265812705c1c9)

File: packages/indexer/src/web3/constants.ts (hunk 5 above)
Comment (melisaguevara) [reply packages/indexer/src/web3/constants.ts:128]: I think for now is ok to return a value as we are using this just to reconstruct the bundles.

**Review by amateima — APPROVED** [review]

File: packages/indexer/src/services/bundles.ts [hunk 11]
```diff
@@ -347,3 +365,133 @@ async function assignBundleExecutedStatus(
...
+  hubClientFactory: utils.HubPoolClientFactory,
+  spokeClientFactory: utils.SpokePoolClientFactory,
+  logger: winston.Logger,
+): Promise<void> {
+  const executedBundles =
+    await bundleRepo.getExecutedBundlesWithoutEventsAssociated({
+      fromBlock: utils.ACROSS_V3_MAINNET_DEPLOYMENT_BLOCK,
+    });
+
+  if (executedBundles.length > 0) {
```
Comment (daywiss) [inline packages/indexer/src/services/bundles.ts:388]: just return early rather than nesting the rest of the logic

**Review by daywiss — APPROVED** [review]
cant say my approval here has much weight

**Pull request merged**